static/js/authController.ts

```diff
@@ -122,6 +122,63 @@ export class AuthController {
     return this.token ? this.token : null;
   }
 
+  public signInWithGoogle(): void {
+    const googleAuth = new firebase.auth.GoogleAuthProvider();
+    
+    firebase.auth().signInWithPopup(googleAuth).catch((error) => {
+      if (error.code === 'auth/account-exists-with-different-credential') {
+        let pendingCredential = error.credential;
+        let email = error.email;
+
+        firebase.auth().fetchSignInMethodsForEmail(email).then((methods: any) => {
+          if (methods.length > 0) {
+            window.alert(`Your auth methods${methods.length}: ${methods[0]}`);
+            if (methods[0] === "facebook.com") {
+              if (window.confirm('You have already authorized using a Facebook account. Click yes if you\'d like to sign in with Facebook and click no if you wish to sign in with a Google account and link to the existing one.')) {
+                this.signInWithFacebook();
+              } else {
+                firebase.auth().signInWithPopup(googleAuth).then((result) => {
+                  result.user!.linkAndRetrieveDataWithCredential(pendingCredential).then(function(usercred) {
+                    // Google account successfully linked to the existing Firebase user.
+                    window.alert('Your Google account has been successfully linked!');
+                  });
+                });
+              }
+            }
+          }
+        });
+      }
+    });
+  }
+
+  public signInWithFacebook(): void {
+    const facebookAuth = new firebase.auth.FacebookAuthProvider();
+    firebase.auth().signInWithPopup(facebookAuth).catch((error) => {
+      if (error.code === 'auth/account-exists-with-different-credential') {
+        let pendingCredential = error.credential;
+        let email = error.email;
+
+        firebase.auth().fetchSignInMethodsForEmail(email).then((methods: any) => {
+          if (methods.length > 0) {
+            window.alert(`Your auth methods${methods.length}: ${methods[0]}`);
+            if (methods[0] === "google.com") {
+              if (window.confirm('You have already authorized using a Facebook account. Click yes if you\'d like to sign in with Facebook and click no if you wish to sign in with a Google account and link to the existing one.')) {
```

Comment:
this can be extracted into a config file

Reply:
There's also probably an argument for extracting the `fetchSignInMethodsForEmail` call into its own method with params